services/motion/builtin/builtin.go

```diff
@@ -260,283 +279,60 @@ func (ms *builtIn) MoveOnGlobe(
 		return false, err
 	}
 
-	// successChan is for sending messages that relate to the base meeting the success criteria for the request
-	successChan := make(chan bool, 1)
-	defer close(successChan)
-
-	// replanChan is for sending messages that signal that we should stop motion and replan
-	// initialize with a true signal so that we immediately start planning
-	replanChan := make(chan bool, 1)
-	defer close(replanChan)
-	replanChan <- true
-
-	// errChan is for sending error messages encountered during the move
-	errChan := make(chan error)
-	defer close(errChan)
-
-	// create a waitgroup to manage goroutines and wait on their completion before exiting the function
-	var backgroundWorkers sync.WaitGroup
-	defer backgroundWorkers.Wait()
-
-	// create a cancellable context and ensure that it is cancelled before exiting
-	var cancelCtx context.Context
-	var cancelFn context.CancelFunc
-	defer func() {
-		if cancelFn != nil {
-			cancelFn()
-		}
-	}()
-
-	// helper function to manage polling functions
-	startPolling := func(ctx context.Context, period time.Duration, fn func(context.Context) error) {
-		backgroundWorkers.Add(1)
-		goutils.ManagedGo(func() {
-			ticker := time.NewTicker(period)
-			defer ticker.Stop()
-			for {
-				select {
-				case <-ctx.Done():
-					return
-				case <-ticker.C:
-					if err := fn(ctx); err != nil {
-						errChan <- err
-					}
-					return
-				}
-			}
-		}, backgroundWorkers.Done)
+	plan, err := plan(ctx, planRequest, kb, componentName.Name)
+	if err != nil {
+		return false, err
 	}
 
+	ps := newPlanSession(ctx, plan)
+	// TODO: Remove replanCount when read position & obsticle polling is integrated
+	var replanCount int
+	ps.start(motionCfg, kb, ms, movementSensor, destination, replanCount)
+
 	// start the loop that (re)plans when something is read from the replan channel
 	// and exits when something is read from the success channel
 	for {
+		replanCount++
+		if err := ctx.Err(); err != nil {
```

Comment:
This fixes a bug where the select statement might not go down the ctx.Done() path if there are messages on the other channels. This change ensures that we always terminate the loop if the context has been cancelled.